Feature/mongodb atlas

Danielschuetzle opened the pull request:
Set up a MongoDB Atlas database and configured the connection.
Implemented the integration with MongoDB Atlas to store new travel plans.
Modified the update functionality to update existing plans in MongoDB Atlas.
Retrieved and displayed travel plans from MongoDB Atlas in the application.
Additionally, the following changes have been made:

- Configured the project to connect to MongoDB Atlas using the provided MONGODB_URI.
- Updated the API routes (pages/api/plans.js) to interact with the MongoDB database.
- Utilized the mongodb package to establish the connection and perform database operations.
- Added support for fetching travel plan data from the database by updating the TravelPlanDetail component and utilizing the getTravelPlan function from the travelPlanDb module.
- Refactored the TravelPlanForm component to implement form submission functionality and save travel plan data to the MongoDB database using the /api/travelPlan endpoint.
- Implemented the redirection of users to the detail page of the newly created plan after submission.
- Created an API route (pages/api/plans.js) to handle fetching and storing travel plans in the MongoDB database.

FelineHuhn commented on lib/travelPlanDb.js:
```diff
@@ -0,0 +1,63 @@
+export const deleteTravelPlan = async (id) => {
+  const db = await connectToDatabase(process.env.MONGODB_URI);
+
+  const result = await db
+    .db(dbName)
+    .collection("travelPlans")
+    .deleteOne({ _id: ObjectId(id) });
```
```suggestion
    const result = await db.collection(collectionName).deleteOne({ _id: ObjectId(id) });

```

FelineHuhn commented on lib/travelPlanDb.js:
```diff
@@ -0,0 +1,63 @@
+  const db = await connectToDatabase(process.env.MONGODB_URI);
+
+  // Insert the plan object into the 'travelPlans' collection in your MongoDB Atlas database
+  const result = await db.db(dbName).collection("travelPlans").insertOne(plan);
```
```suggestion
  const result = await db.collection(collectionName).insertOne(plan);
```

FelineHuhn commented on pages/Plan/[id].js:
```diff
@@ -1,16 +1,26 @@
 import { useRouter } from 'next/router';
```
```suggestion
```

FelineHuhn replied:
Since you don't use the useRouter in this file, the import should be removed here.

FelineHuhn commented on pages/Plan/[id].js:
```diff
@@ -1,16 +1,26 @@
+
+export async function getServerSideProps(context) {
+  const { id } = context.params;
+  const res = await fetch(`http://localhost:3000/api/plans/${id}`);
+  const data = await res.json();
+
+  return {
+    props: {
+      plan: data
+    }
+  };
```
```suggestion
try {
    const res = await fetch(`http://localhost:3000/api/plans/${id}`);
    const data = await res.json();

    return {
      props: {
        plan: data,
      },
    };
  } catch (error) {
    console.error('Error retrieving travel plan:', error);
    return {
      props: {
        plan: null,
      },
    };
  }
```

FelineHuhn replied:
An error handling would be great here, to catch errors.

FelineHuhn commented on components/TravelPlansList/index.js:
```diff
@@ -0,0 +1,37 @@
+  return data;
+};
+
+const TravelPlansList = ({ travelPlans }) => {
```
```suggestion
const TravelPlansList = () => {
```

FelineHuhn replied:
The `travelPlans` prop could be removed, since you now fetch the data with `useSWR`, or?

FelineHuhn commented on components/TravelPlansList/index.js:
```diff
@@ -0,0 +1,37 @@
+    return <div>Error: {error.message}</div>;
+  }
+
+  const plans = fetchedPlans || travelPlans;
```
```suggestion
  const plans = fetchedPlans
```

FelineHuhn replied:
See above

FelineHuhn commented on components/TravelPlanForm/index.js:
```diff
@@ -1,47 +1,56 @@
+  const router = useRouter();
+
+  // Initialize the state of the form fields
   const [plan, setPlan] = useState({
```
One suggestion: You could put the object in an initialPlanState variable for better readablity.

FelineHuhn replied:
```suggestion
  const [plan, setPlan] = useState(initialPlanState);
```

FelineHuhn commented on components/TravelPlanForm/index.js:
```diff
@@ -1,47 +1,56 @@
 
 const TravelPlanForm = ({ addTravelPlan }) => {
+  const router = useRouter();
+
```
```suggestion
 const initialPlanState = {
    name: '',
    startDate: '',
    endDate: '',
    activity: '',
  };
```

FelineHuhn commented on components/TravelPlanForm/index.js:
```diff
@@ -1,47 +1,56 @@
     startDate: '',
     endDate: '',
-    activity: ''
+    activity: '',
```
Than you can remove line 9 to 12.

FelineHuhn commented on components/TravelPlanForm/index.js:
```diff
@@ -1,47 +1,56 @@
+    router.push(`/plan/${result._id}`);
+
+    // Clear the form fields
     setPlan({
```
```suggestion
    setPlan(initialPlanState);
```
And remove the lines below.

FelineHuhn commented on components/TravelPlanDetail/index.js:
```diff
@@ -1,8 +1,32 @@
+  }
+
+  // Sort the activities by their start dates
+  const sortedActivities = plan.activities.sort(
```
```suggestion
  const sortedActivities = [...plan.activities].sort(
```

FelineHuhn replied:
You should create a copy of `plan.activities` before sorting it. This ensures that the original `plan.activities` array is not mutated.

Danielschuetzle commented:
There was an issue where the plans.map function was throwing an error due to plans not being an array in the TravelPlansList component.

To resolve the issue, I made the following changes:

Updated the getPlans function in travelPlanDb.js to fetch all travel plans instead of differentiating between fetching all plans and a specific plan by ID.
Modified the handler function in api/plans.js to use the updated getPlans function to fetch all plans.
With these changes, the TravelPlansList component now receives an array of plans, allowing the map function to be used without errors.

Katemuya commented:
LGTM 🚀

My knowledge in Backend is not the best but applying the little I know, as well as seeing that you have applied Feline's suggestions, I will approve the PR. 
If you need a second opinion please feel free to change the reaction in slack.

cheers 🤗
